File: tests/instruments/tick-watch/codec-scaling.lua
-- Codec-only experiment. No entities, surfaces, transfer jobs or event hooks.
local KEY='surface_export_codec_scaling_lab'
local function equal(a,b)
  if type(a)~=type(b) then return false end
  if type(a)~='table' then return a==b end
  for k,v in pairs(a) do if not equal(v,b[k]) then return false end end
  for k in pairs(b) do if a[k]==nil then return false end end
  return true
end
local function measured(label,fn)
  local p=helpers.create_profiler()
  local value=fn()
  p.stop()
  rcon.print({'','[SE_CODEC_V1]',label,'\t',p})
  return value
end
local invoke
invoke=function(action,id,arg)
  assert(id:match('^codecscale%-%w+$'),'owned ID required')
  local s=package.loaded[KEY]
  if action=='cleanup' then
    if not s then return {success=true,removed=true} end
    assert(s.id==id,'foreign lab state')
    package.loaded[KEY]=nil
    assert(storage.platform_exports[s.export_id].payload==s.cached,'source cache changed')
    return {success=true,removed=true,sourcePayloadUnchanged=true}
  end
  if action=='status' then return {success=true,present=s~=nil} end
  assert(#game.connected_players==0 and not game.tick_paused,'unoccupied running host required')
  assert(table_size(storage.async_jobs or {})==0 and table_size(storage.locked_platforms or {})==0
    and table_size(storage.destination_holds or {})==0,'active transfer work')
  if action=='prepare' then
    assert(not s,'another codec lab exists')
    assert(script.active_mods.base=='2.1.17')
    local cached=assert(storage.platform_exports[arg.export_id],'export missing').payload
    assert(type(cached)=='string' and #cached<200000)
    s={id=id,invoke=invoke,export_id=arg.export_id,cached=cached,ranges=arg.ranges}
    package.loaded[KEY]=s
    s.json=assert(helpers.decode_string(cached));assert(#s.json<=4000000)
    s.expected=assert(helpers.json_to_table(s.json))
    s.metadata={}
    for k,v in pairs(s.expected) do
      if k~='entities' and k~='tiles' and k~='belt_side_groups' then s.metadata[k]=v end
    end
    assert(#s.ranges>0 and #s.ranges<100)
    return {success=true,engine=script.active_mods.base,mods=script.active_mods,bytes=#s.json,framesPerCopy=#s.ranges+1}
  end
  assert(s and s.id==id,'lab absent or foreign')
  if action=='begin' then
    assert(arg.scale==1 or arg.scale==2 or arg.scale==4)
    s.scale=arg.scale;s.result={copies={}};s.cursor=1;s.previous_emit=nil;s.previous_consume=nil
    local copies={};for i=1,s.scale do copies[i]=s.json end
    s.monolith='{"copies":['..table.concat(copies,',')..']}'
    s.compressed=nil
    return {success=true,jsonBytes=#s.monolith,frames=(#s.ranges+1)*s.scale}
  end
  if action=='compress_full' then
    s.compressed=measured('factorio_compress',function() return assert(helpers.encode_string(s.monolith)) end)
    return {success=true,tick=game.tick,encodedBytes=#s.compressed}
  end
  if action=='decode_full' then
    local result=measured('full_decode',function()
      local json=assert(helpers.decode_string(s.compressed))
      return assert(helpers.json_to_table(json))
    end)
    assert(#result.copies==s.scale)
    for _,copy in ipairs(result.copies) do assert(equal(copy,s.expected),'full payload differs') end
    return {success=true,tick=game.tick,exactPayload=true}
  end
  if action=='emit' then
    if s.previous_emit==game.tick then return {success=true,waiting=true} end
    s.previous_emit=game.tick
    assert(arg==s.cursor,'unexpected source sequence')
    local per_copy=#s.ranges+1
    local copy=math.floor((arg-1)/per_copy)+1;local index=(arg-1)%per_copy
    assert(copy<=s.scale)
    local json=measured('source_encode',function()
      local frame={v=1,seq=arg,copy=copy}
      if index==0 then frame.kind='metadata';frame.data=s.metadata
      else
        local range=s.ranges[index];frame.kind='array';frame.key=range.key;frame.first=range.first
        assert(range.key=='entities' or range.key=='tiles' or range.key=='belt_side_groups')
        frame.data={};for i=range.first,range.last do frame.data[#frame.data+1]=assert(s.expected[range.key][i]) end
      end
      return helpers.table_to_json(frame)
    end)
    assert(#json<=65536,'frame size limit exceeded')
    s.cursor=s.cursor+1
    return {success=true,tick=game.tick,seq=arg,json=json,bytes=#json}
  end
  if action=='consume' then
    if s.previous_consume==game.tick then return {success=true,waiting=true} end
    s.previous_consume=game.tick
    assert(type(arg)=='string' and #arg<=100000)
    local frame
    measured('destination_decode',function()
      local json=assert(helpers.decode_string(arg),'external deflate incompatible')
      assert(#json<=65536,'decoded frame oversized')
      frame=assert(helpers.json_to_table(json))
      assert(frame.v==1 and frame.seq==(s.received or 0)+1 and frame.copy<=s.scale)
      if frame.kind=='metadata' then
        assert(s.result.copies[frame.copy]==nil);s.result.copies[frame.copy]=frame.data
      else
        assert(frame.kind=='array');local copy=assert(s.result.copies[frame.copy])
        local target=copy[frame.key] or {};copy[frame.key]=target
        assert(frame.first==#target+1,'missing or duplicated array range')
        for _,v in ipairs(frame.data) do target[#target+1]=v end
      end
      s.received=frame.seq
    end)
    return {success=true,tick=game.tick,seq=frame.seq}
  end
  if action=='reset_receiver' then s.result={copies={}};s.received=0;return {success=true} end
  if action=='verify' then
    assert(s.received==(#s.ranges+1)*s.scale and #s.result.copies==s.scale)
    for _,copy in ipairs(s.result.copies) do assert(equal(copy,s.expected),'sectional payload differs') end
    return {success=true,exactPayload=true,copies=s.scale}
  end
  if action=='gzip_shape' then
    local ok,value=pcall(helpers.decode_string,arg)
    return {success=true,accepted=ok and value=='codec-shape',returned=type(value),error=not ok and tostring(value) or nil}
  end
  error('unknown action')
end
return invoke

File: tests/instruments/tick-watch/codec-scaling.mjs
import assert from 'node:assert/strict';
import {readFileSync,writeFileSync} from 'node:fs';
import {spawn} from 'node:child_process';
import {createHash} from 'node:crypto';
import {inflateSync,gunzipSync} from 'node:zlib';
import {withWorkflowLock} from '../../../tools/shared/workflow-lock.mjs';
import {lua,preflightState,assertLeaseClean} from '../../lab-gallery/batch-lifecycle.mjs';

const mode=process.argv[2];
assert.ok(['--run','--cleanup-proof','--analyze','--recorded','--cleanup'].includes(mode));
const hash=bytes=>createHash('sha256').update(bytes).digest('hex');
function analyze(r) {
  assert.equal(r.verdict,'PASS',r.error);assert.equal(r.cleanup.removed,true);assert.equal(r.residue.present,false);
  assert.deepEqual(r.cases.map(c=>c.scale),[1,2,4]);
  return r.cases.flatMap(c=>{assert.equal(c.trials.length,2);return c.trials.map(t=>{
    assert.equal(t.verification.exactPayload,true);assert.equal(t.control.exactPayload,true);
    for(const side of ['source','destination']) {
      assert.equal(t[side].length,t.frames);
      for(let i=0;i<t[side].length;i++) {
        assert.equal(t[side][i].seq,i+1);
        assert.ok(t[side][i].readings.every(v=>Number.isFinite(v.ms)&&v.ms>=0));
        if(i)assert.ok(t[side][i].tick>t[side][i-1].tick);
      }
    }
    return {scale:c.scale,repeat:t.repeat,frames:t.frames,MB:t.jsonBytes/1e6,
      fullCompressMs:t.control.compressionMs,fullDecodeMs:t.control.decodeMs,
      maxEncodeMs:t.maxSourceEncodeMs,maxDecodeMs:t.maxDestinationDecodeMs,
      maxSourceActionMs:t.maxSourceActionMs,maxDestinationActionMs:t.maxDestinationActionMs,
      fullBytes:t.control.encodedBytes,sectionBytes:t.encodedBytes,nodeCompressAwaitMs:t.nodeCompressionAwaitMs,
      jsonlGzipBytes:t.jsonlGzip.gzipBytes};
  });});
}
if(mode==='--recorded') {
  const root=new URL('./evidence/',import.meta.url);
  const manifest=JSON.parse(readFileSync(new URL('codec-2.1.17.manifest.json',root),'utf8'));
  const packed=readFileSync(new URL('codec-2.1.17.json.gz',root));
  assert.equal(hash(packed),manifest.sha256);
  const raw=gunzipSync(packed);assert.equal(hash(raw),manifest.decodedSha256);
  const bundle=JSON.parse(raw);assert.equal(bundle.version,1);
  const reports=bundle.reports.map(text=>JSON.parse(text));
  assert.deepEqual(reports.map(r=>r.id),manifest.reportIds);
  for(const r of reports)for(const [key,value] of Object.entries({...bundle.sources,payload:bundle.payload,api:bundle.api})) {
    assert.equal(hash(value),r.hashes[key],`${r.id}: ${key} provenance mismatch`);
  }
  const [proof,matrix]=reports;
  assert.equal(proof.cleanupProofPassed,true);assert.equal(proof.cleanup.removed,true);
  assert.equal(proof.residue.present,false);assert.equal(proof.postflight,'PASS');
  assert.equal(matrix.postflight,'PASS');
  console.table(analyze(matrix));
  console.log('Recorded evidence hashes, original harness, cleanup proof and six codec trials verified. No cluster calls.');
} else if(mode==='--analyze')console.table(analyze(JSON.parse(readFileSync(process.argv[3],'utf8'))));
else await withWorkflowLock(async()=>{
  if(mode==='--cleanup') {
    const id=process.argv[3];assert.match(id,/^codecscale-[a-z0-9]+$/);
    const result=lua(2,`local s=package.loaded.surface_export_codec_scaling_lab;if not s then return {success=true,removed=true} end;return s.invoke('cleanup','${id}')`);
    assert.equal(result.success,true,result.error);console.log(result);return;
  }
  const id=`codecscale-${Date.now().toString(36)}`;
  const exportId=process.argv[3] || '178_transfer-cleanup-tickwatch-mtuop1r6';
  assert.match(exportId,/^[A-Za-z0-9_:.-]+$/,'invalid export cache ID');
  const raw=readFileSync(process.argv[4] || 'ci-artifacts/import-setup-exact-payload.json'),payload=JSON.parse(raw);
  const luaCode=readFileSync(new URL('./codec-scaling.lua',import.meta.url),'utf8');
  const nodeCode=readFileSync(new URL('./codec-scaling.cjs',import.meta.url),'utf8');
  const apiRaw=readFileSync(process.argv[5] || 'ci-artifacts/issue-69-runtime-api.json');const api=JSON.parse(apiRaw);
  assert.equal(api.application_version,'2.1.17');
  const members={LuaHelpers:['create_profiler','decode_string','encode_string','json_to_table','table_to_json'],
    LuaProfiler:['stop'],LuaGameScript:['connected_players','tick_paused','tick'],LuaBootstrap:['active_mods'],LuaRCON:['print']};
  for(const [name,required] of Object.entries(members)) {
    const cls=api.classes.find(c=>c.name===name);assert.ok(cls,`missing API class ${name}`);
    for(const member of required)assert.ok([...cls.methods,...cls.attributes].some(m=>m.name===member),`missing API ${name}.${member}`);
  }
  const ranges=[];
  for(const key of ['entities','tiles','belt_side_groups']) {
    const values=payload[key];assert.ok(Array.isArray(values));let first=0,bytes=150;
    for(let i=0;i<values.length;i++) {
      const size=Buffer.byteLength(JSON.stringify(values[i]))+1;assert.ok(size<59000);
      if(bytes+size>59000&&i>first){ranges.push({key,first:first+1,last:i});first=i;bytes=150;}
      bytes+=size;
    }
    ranges.push({key,first:first+1,last:values.length});
  }
  const input={id,exportId,ranges,lua:luaCode,cleanupProof:mode==='--cleanup-proof',
    hashes:{lua:hash(luaCode),node:hash(nodeCode),runner:hash(readFileSync(new URL(import.meta.url))),payload:hash(raw),api:hash(apiRaw)},
    contract:{engine:'2.1.17',scales:[1,2,4],repeats:2,scaleDefinition:'Copies of the captured 3.09 MB payload in a codec container, not physically larger platforms.',
      invariant:'All fields and values in each decoded copy equal the captured source; sequence/range checks; archive text round trips exactly.',
      route:'Both Factorio codec endpoints on unoccupied host-2; external async compression in Node in that container; direct persistent localhost RCON.',
      bounds:'At most 100 sections per copy, 4 copies, 2 trials per scale, 2200 requests, 360s loop deadline, 30s RCON timeout, 100000-byte requests, 300000-character responses. Native frame <=65536 bytes.',
      mutation:'Temporary package state only; no surfaces/entities/jobs/event hooks/save/config changes. Original export cache retained.',
      timing:'Phase profilers plus RCON action/reply-construction profiler. Compilation and socket wait excluded from Lua readings; recorded RCON RTT separate. Oracle/setup actions are lab overhead.',
      limits:'Preparation, monolithic JSON assembly, source capture, network transport across instances, production recovery and client FPS/UPS are not acceptance-tested here. Deferred garbage collection is not isolated.'}};
  assertLeaseClean(2,preflightState(2),'codec scaling preflight');
  const cached=lua(2,`local e=assert(storage.platform_exports['${exportId}']);return {success=true,payload=e.payload}`);
  assert.equal(cached.success,true,cached.error);assert.equal(hash(inflateSync(Buffer.from(cached.payload,'base64'))),hash(raw));
  const report=await new Promise((resolve,reject)=>{
    const child=spawn('docker',['exec','-i','surface-export-host-2','node','-'],{stdio:['pipe','pipe','pipe']});
    let pending='',result,stderr='';
    const timer=setTimeout(()=>{child.kill();reject(Error(`Outer deadline exceeded. Verify stopped runner, then use --cleanup ${id}; do not assume removal.`));},450000);
    child.stdout.on('data',chunk=>{
      pending+=chunk;const lines=pending.split('\n');pending=lines.pop();
      for(const line of lines)if(line.trim())try{const event=JSON.parse(line);if(event.event==='report')result=event.report;else console.log(event);}catch(error){console.warn("Probe emitted a non-JSON line:",error.message);stderr+=line.slice(0,500);}
    });
    child.stderr.on('data',chunk=>{stderr+=String(chunk).slice(0,1000);});
    child.on('error',error=>{clearTimeout(timer);reject(error);});
    child.on('close',exit=>{clearTimeout(timer);result?resolve(result):reject(Error(`Codec runner exited ${exit}: ${stderr}`));});
    child.stdin.end(nodeCode+`\nrun(${JSON.stringify(input)}).catch(e=>{console.error(e.message);process.exitCode=1;});\n`);
  });
  try{assertLeaseClean(2,preflightState(2),'codec scaling postflight');report.postflight='PASS';}
  catch(error){report.postflightError=error.message;report.verdict='HARNESS_ERROR';}
  const file=`ci-artifacts/${id}.json`;writeFileSync(file,JSON.stringify(report,null,2));
  console.log({artifact:file,verdict:report.verdict,error:report.error,cleanup:report.cleanup,cleanupProofPassed:report.cleanupProofPassed});
  if(report.verdict==='PASS')console.table(analyze(report));
  else if(!report.cleanupProofPassed||report.postflightError)process.exitCode=1;
});
